### PR TITLE
SPARK-738: Spark should detect and wrap nonserializable exceptions

### PR DESCRIPTION
**Pull request by pwendell**

### CODE REVIEW

**Comment by pwendell** [comment]
Jenkins, test this please.

**Comment by mridulm** [comment]
Since Throwable is Serializable (and so all Error's and Exception's are
too), the root cause here are buggy extensions : we should be fixing that
if it is from within spark.
This hack should be required only when what is thrown is outside of our
codebase, and cant influence a fix.

Regards,
Mridul

**Comment by rxin** [comment]
Yes this should catch user generated exceptions that are not serializable.

File: core/src/main/scala/spark/executor/Executor.scala [hunk 1]
```diff
@@ -123,7 +123,19 @@ private[spark] class Executor(executorId: String, slaveHostname: String, propert
 
         case t: Throwable => {
           val reason = ExceptionFailure(t)
-          context.statusUpdate(taskId, TaskState.FAILED, ser.serialize(reason))
+          val serReason =
+            try {
```
Comment (rxin) [inline core/src/main/scala/spark/executor/Executor.scala]: put catch on the same line as the closing bracket

} catch {

also don't use abbreviated "exn" in the exception message.

File: core/src/main/scala/spark/executor/Executor.scala [hunk 2]
```diff
@@ -1,6 +1,6 @@
 package spark.executor
 
-import java.io.{File, FileOutputStream}
+import java.io.{NotSerializableException, File, FileOutputStream}
```
Comment (rxin) [inline core/src/main/scala/spark/executor/Executor.scala]: Put NotSerializableException after File / FileOutputStream

**Comment by mateiz** [comment]
Patrick, as a high-level approach, why don't we just send the exception's message and stack trace (array of StackTraceElement) instead of the exception itself? I'd prefer not to have a special case for this.

**Comment by mateiz** [comment]
(At the end of the day, those are the only pieces of information we need to display.)

**Comment by rxin** [comment]
That's an excellent idea - we should return the exception class name, message, and stack trace.

**Comment by pwendell** [comment]
I agree - let me refactor this and post tomorrow.

**Comment by pwendell** [comment]
Okay I changed it to use the new approach. I didn't include classname (it is not used downstream and there are other fields in exception we could include also - so I figured keep it minimal for now).

File: core/src/main/scala/spark/TaskEndReason.scala [hunk 3]
```diff
@@ -14,9 +14,16 @@ private[spark] case object Success extends TaskEndReason
 private[spark] 
 case object Resubmitted extends TaskEndReason // Task was finished earlier but we've now lost it
 
-private[spark] 
-case class FetchFailed(bmAddress: BlockManagerId, shuffleId: Int, mapId: Int, reduceId: Int) extends TaskEndReason
+private[spark] case class FetchFailed(
```
Comment (pwendell) [inline core/src/main/scala/spark/TaskEndReason.scala:17]: This is code cleanup - it exceeded 100 characters before.

**Comment by mridulm** [comment]
Can you please also include the exception/error classname too ?
Particularly with generated code, the exception name is more useful than the stack trace (and reverse in bunch of other cases; unfortunately).

Thanks !
Mridul

**Comment by pwendell** [comment]
Mridul - do you mean exn.getClass()? I thought most exceptions report their own type in toString(), could you give an example where it's not sufficient?

scala> val x = new IOException("test")
x: java.io.IOException = java.io.IOException: test

scala> x.toString()
res4: java.lang.String = java.io.IOException: test

scala> x.getClass().toString()
res5: java.lang.String = class java.io.IOException

scala> class MyException extends Exception
defined class MyException

scala> val f = new MyException()
f: MyException = MyException

scala> f.toString()
res6: java.lang.String = MyException

**Comment by mridulm** [comment]
This is just a convention; and not mandated by the api contract. (Unlike
Throwable implements Serializable - which, as we see, is also being
violated !).
All we need is just exn.getClass().getName() - not necessary to serialize
the entire class across ofcourse.

Thanks,
Mridul

**Comment by pwendell** [comment]
Okay - since Mridul and Reynold both seem to support - I added the exception class name.

**Comment by pwendell** [comment]
Jenkins, test this please (just trying to debug something with Matt, this still isn't working for me)

**Comment by mateiz** [comment]
Thanks Patrick!

**Pull request merged**